engines/registry/app/controllers/registry/registry_controller.rb

```diff
@@ -0,0 +1,100 @@
+      render json: { token: token }, status: :ok
+    end
+
+    # Catalog handler
+    # Returns a Distribution Registry HTTP API V2 - compatible repository catalog as defined in
+    # https://distribution.github.io/distribution/spec/api/#listing-repositories
+    def catalog
+      access_scope = Registry::AccessScope.parse(['registry:catalog:*'])
+      repos = access_scope.allowed_paths(System.find_by(login: @client&.account))
+      logger.debug("Returning #{repos.size} repos for client #{@client}")
+
+      response.set_header('Docker-Distribution-Api-Version', REGISTRY_API_VERSION)
+      render json: { repositories: repos }, status: :ok
+    end
+
+    private
+
+    # Support multiple scopes. Podman & Docker handle this differently
+    #   - Podman sends multiple querystrings with the same name.
+    #   - Docker sends space-separated values in only one query string.
+    # This should normalize everything
+    def set_requested_scopes
+      raw_scopes = CGI.parse(request.env['QUERY_STRING']).fetch('scope', [])
+      .join(' ')
+      .split
+      .map(&:presence)
+      .compact
+      .map(&:downcase)
+
+      @requested_scopes = []
+      @requested_scopes = raw_scopes.map { |scope| Registry::AccessScope.parse(scope) } unless raw_scopes.empty?
+
+      logger.info("Requested scopes: #{@requested_scopes.map(&:to_s)}")
+    end
+
+    # AuthN handler
+    # https://docs.docker.com/registry/spec/auth/jwt/#getting-a-bearer-token
+    def basic_auth
+      # skip authentication if this is not a login request
+      return unless request.authorization
+
+      authenticate_or_request_with_http_basic('SUSE Registry Authentication') do |login, password|
+        begin
+          @client = Registry::AuthenticatedClient.new(login, password)
+        rescue StandardError
+          logger.info _('Could not find system with login \"%{login}\" and password \"%{password}\"') %
+            { login: login, password: password }
+          error = ActionController::TranslatedError.new(N_('Registration expired, please re authenticate again'))
+          error.status = :unauthorized
+          raise error
+        end
+
+        true
+      end
+    end
+
+    def catalog_token_auth
+      authenticate_or_request_with_http_token(authorize_api_registry_url, 'authentication required') do |token|
+        begin
+          @client = Registry::CatalogClient.new(token)
+        rescue JWT::DecodeError
+          logger.info _('Invalid token')
+          error = ActionController::TranslatedError.new(N_('Invalid registry token'))
+          error.status = :unauthorized
+          raise error
+        end
+
+        @client.authorized_for_catalog?
+      end
+    end
+
+    # is called by authenticate_or_request_with_http_token when client provides no token
+    def request_http_token_authentication(realm = authorize_api_registry_url, message = 'authentication required')
```

Comment:
Please have a look at the most recent version of [glue/app/controllers/api/registry_controller.rb](https://github.com/SUSE/happy-customer/blob/3f818723ea1b8cceab28e01505ea13d0570e28e1/glue/app/controllers/api/registry_controller.rb#L122
)
We had to add a fix here that broke some registry clients.